jax_distribution/systems/ff_mappo_rware.py

```diff
@@ -0,0 +1,508 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import copy
+import os
+from logging import Logger as SacredLogger
+from os.path import abspath, dirname
+from typing import Any, Callable, Dict, Sequence, Tuple
+
+import chex
+import distrax
+import flax.linen as nn
+import hydra
+import jax
+import jax.numpy as jnp
+import jumanji
+import numpy as np
+import optax
+from colorama import Fore, Style
+from flax.core.frozen_dict import FrozenDict
+from flax.linen.initializers import constant, orthogonal
+from jumanji.env import Environment
+from jumanji.environments.routing.robot_warehouse.generator import RandomGenerator
+from jumanji.types import Observation
+from jumanji.wrappers import AutoResetWrapper
+from omegaconf import DictConfig, OmegaConf
+from optax._src.base import OptState
+from sacred import Experiment, observers, run, utils
+
+from jax_distribution.evaluator import evaluator_setup
+from jax_distribution.logger import logger_setup
+from jax_distribution.types import ExperimentOutput, LearnerState, PPOTransition
+from jax_distribution.utils.jax import merge_leading_dims
+from jax_distribution.utils.logger_tools import config_copy, get_logger
+from jax_distribution.utils.timing_utils import TimeIt
+from jax_distribution.wrappers.jumanji import (
+    LogWrapper,
+    ObservationGlobalState,
+    RwareMultiAgentWithGlobalStateWrapper,
+)
+
+
+class ActorCritic(nn.Module):
+    """Actor Critic Network."""
+
+    action_dim: Sequence[int]
+
+    @nn.compact
+    def __call__(self, observation: Observation) -> Tuple[distrax.Categorical, chex.Array]:
+        """Forward pass."""
+        x = observation.agents_view
+
+        actor_output = nn.Dense(64, kernel_init=orthogonal(np.sqrt(2)), bias_init=constant(0.0))(x)
+        actor_output = nn.relu(actor_output)
+        actor_output = nn.Dense(64, kernel_init=orthogonal(np.sqrt(2)), bias_init=constant(0.0))(
+            actor_output
+        )
+        actor_output = nn.relu(actor_output)
+        actor_output = nn.Dense(
+            self.action_dim, kernel_init=orthogonal(0.01), bias_init=constant(0.0)
+        )(actor_output)
+
+        masked_logits = jnp.where(
+            observation.action_mask,
+            actor_output,
+            jnp.finfo(jnp.float32).min,
+        )
+        actor_policy = distrax.Categorical(logits=masked_logits)
+
+        y = observation.global_state
+
+        critic_output = nn.Dense(64, kernel_init=orthogonal(np.sqrt(2)), bias_init=constant(0.0))(y)
+        critic_output = nn.relu(critic_output)
+        critic_output = nn.Dense(64, kernel_init=orthogonal(np.sqrt(2)), bias_init=constant(0.0))(
+            critic_output
+        )
+        critic_output = nn.relu(critic_output)
+        critic_output = nn.Dense(1, kernel_init=orthogonal(1.0), bias_init=constant(0.0))(
+            critic_output
+        )
+
+        return actor_policy, jnp.squeeze(critic_output, axis=-1)
```

Comment:
Interesting this uses the same network? I would have though we'd only need to call the critic once per all agents?

Reply:
We used out_axes  None for the critic to ignore/pass the multiple calculation